crates/subspace-proof-of-time/src/lib.rs

```diff
@@ -0,0 +1,93 @@
+//! Proof of time implementation.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+mod pot_aes;
+
+use subspace_core_primitives::{BlockHash, PotKey, PotProof, PotSeed, SlotNumber};
+
+#[derive(Debug)]
+#[cfg_attr(feature = "thiserror", derive(thiserror::Error))]
+pub enum PotVerificationError {
+    #[cfg_attr(
+        feature = "thiserror",
+        error("Unexpected number of checkpoints: {expected}, {actual}")
+    )]
+    CheckpointCountMismatch { expected: u8, actual: u64 },
+
+    #[cfg_attr(feature = "thiserror", error("Checkpoint verification failed"))]
+    VerificationFailed,
+}
+
+/// Wrapper for the low level AES primitives
+pub struct ProofOfTime {
+    /// Number of checkpoints per PoT.
+    num_checkpoints: u8,
+
+    /// Number of chained AES operations per checkpoint.
+    checkpoint_iterations: u32,
+}
+
+impl ProofOfTime {
+    /// Creates the AES wrapper.
+    pub fn new(num_checkpoints: u8, checkpoint_iterations: u32) -> Self {
+        Self {
+            num_checkpoints,
+            checkpoint_iterations,
+        }
+    }
+
+    /// Builds the proof.
+    pub fn create(
+        &self,
+        seed: PotSeed,
+        key: PotKey,
+        slot_number: SlotNumber,
+        injected_block_hash: BlockHash,
+    ) -> PotProof {
+        PotProof::new(
+            slot_number,
+            seed,
+            key,
+            pot_aes::create(
+                &seed,
+                &key,
+                self.num_checkpoints,
+                self.checkpoint_iterations,
+            ),
+            injected_block_hash,
+        )
+    }
+
+    /// Verifies the proof.
+    pub fn verify(&self, proof: &PotProof) -> Result<(), PotVerificationError> {
+        // TODO: this check may break upgrades, revisit.
```

Comment:
Break in which way? You mean it doesn't support dynamic updates or what?

Reply:
This scenario I had was: one side is upgraded to have different number of checkpoints per proof, other side still has the old settings. This needs more thought, should probably go into the spec itself